### PR TITLE
feat: add magic completion for expr.match

### PR DESCRIPTION
**Pull request by stormckey**
This pull request implements the magic completion for match first mentioned in #8679.

Here is the effect:
image
image

### CODE REVIEW

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by magnus-madsen** [comment]
@stormckey Nice. I will take a look soon. Meanwhile, can you add your name to `AUTHORS.md`? Thanks!

File: main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala [hunk 1]
```diff
@@ -66,7 +66,8 @@ object CompletionProvider {
       // Expressions.
       //
       case SyntacticContext.Expr.Constraint => PredicateCompleter.getCompletions(ctx) ++ KeywordCompleter.getConstraintKeywords
-      case SyntacticContext.Expr.InvokeMethod(tpe, name) => InvokeMethodCompleter.getCompletions(tpe, name)
+      case SyntacticContext.Expr.InvokeMethod(tpe, name) =>
+        MagicMatchCompleter.getCompletions(name.name, tpe, ctx) ++ InvokeMethodCompleter.getCompletions(tpe, name)
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala:70]: (Relying on this specific error will be much more robust)

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala:70]: (One concern with my suggestion would be whether it makes it hard to construct the right completion/location. You can experiment and report back if this does not work.)

[the remark this answers is not in the document]
Comment (xxxxxxxyyyyyyyxxxxx) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala:70]: But line 125 seems to have nothing to do with UndefinedField? And I could only find UndefinedStructField and UndefinedJvmStaticField, which seems irreverent to enum🤔

[the remark this answers is not in the document]
Comment (stormckey) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala:70]: But line 125 seems to have nothing to do with UndefinedField? And I could only find UndefinedStructField and UndefinedJvmStaticField, which seems irreverent to enum🤔

**Comment by stormckey** [comment]
> Another test cases that would be useful to cover:
> 
> ```
> enum Foo {
>     case Bar,
>     case Baz,
>     case Quix(String)
> }
> 
> def f(): Foo = Foo.Bar
> 
> def b(www: Foo): Unit = f().ma
> ```

This is a bit tricky, since f() is not included in the range of the completion context, maybe we should consider manually crafting a TextEdit?

File: main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala:70]: I just meant to add it there. In the getSemanticCompletions

**Comment by magnus-madsen** [comment]
> > Another test cases that would be useful to cover:
> > ```
> > enum Foo {
> >     case Bar,
> >     case Baz,
> >     case Quix(String)
> > }
> > 
> > def f(): Foo = Foo.Bar
> > 
> > def b(www: Foo): Unit = f().ma
> > ```
> 
> This is a bit tricky, since f() is not included in the range of the completion context, maybe we should consider manually crafting a TextEdit?

Yes. This is tricky. I am not sure if we can do it, but I would like for us to try, and see if it possible.

**Comment by magnus-madsen** [comment]
@stormckey Would it make sense for you to fix  the small things (e.g. copyright header) and then we merge this, and you can iterate on the rest of the comments?

**Comment by stormckey** [comment]
> @stormckey Would it make sense for you to fix the small things (e.g. copyright header) and then we merge this, and you can iterate on the rest of the comments?

Certainly, I believe I’ve already addressed the minor issues with an additional commit. Should we open a new issue to discuss the next steps?

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
@stormckey Yes, just add yourself to AUTHORS.md and then we can iterate.

I will open a new ticket.

**Comment by stormckey** [comment]
Done

**Comment by magnus-madsen** [comment]
Congratulations on landing your first PR!

**Pull request merged**